torch/_inductor/memory.py

```diff
@@ -0,0 +1,555 @@
+        (i) if there are nodes whose mem1 values are below the max memory so far,
+            then pick the node with the lowest mem2 value;
+        (ii) otherwise, pick the one with the lowest mem1 value.
+    """
+
+    # compute nodes' number of unmet dependencies (for schedulability)
+    # initialize the list of nodes ready to be scheduled
+    nodes_to_schedule: Set[BaseSchedulerNode] = set()
+    for node in nodes:
+        # note that .unmet_dependencies could have deps with the same name
+        # and in that case, it should only be counted once
+        node.mpi_node.indegree = len(node.mpi_node.pred_nodes)
+        if node.mpi_node.indegree == 0:
+            nodes_to_schedule.add(node)
+
+    # compute buffers' number of unmet successors (used to decide when to free)
+    for buf in list(name_to_buf.values()) + list(name_to_input_buf.values()):
+        buf.mpi_buffer.outdegree = len(buf.mpi_buffer.succ_nodes)
+        if buf.get_name() in graph_outputs:
+            buf.mpi_buffer.outdegree += 1
+
+    # initialize memory estimations
+    live_memory = sum(
+        input_buf.mpi_buffer.size_free for input_buf in name_to_input_buf.values()
+    )
+
+    # this is the total output memory, which is a lower bound for peak memory
+    output_memory = sum(
+        name_to_buf[buf_name].mpi_buffer.size_free
+        for buf_name in graph_outputs
+        if buf_name in name_to_buf
+    )
+    max_memory = max(live_memory, output_memory)
+
+    # compute the amount of memory that is allocated when a node is scheduled
+    # and the amount of memory that can be freed when a node is scheduled
+    for i, node in enumerate(nodes):
+        node.mpi_node.index = i  # keep track of the original order
+        node.mpi_node.size = sum(
+            buffer.mpi_buffer.size_alloc for buffer in node.get_outputs()
+        )
+        node.mpi_node.memory_to_free = 0
+        # 1. if a buffer read by this node is last used by this node
+        #    then the buffer can be freed
+        for buf in node.mpi_node.pred_buffers:
+            if buf.mpi_buffer.outdegree == 1:
+                node.mpi_node.memory_to_free += buf.mpi_buffer.size_free
+        # 2. if a buffer written by this node is used internally and
+        #    not needed afterwards, it can be freed
+        for buf in node.get_outputs():
+            if buf.mpi_buffer.outdegree == 0:
+                node.mpi_node.memory_to_free += buf.mpi_buffer.size_free
+
+    # schedule nodes one at a time
+    schedule: List[BaseSchedulerNode] = []
+    num_iters: int = 0
+    while num_iters < len(nodes) and nodes_to_schedule:
+        # select a node to schedule:
+        selected_node = min(
+            nodes_to_schedule,
+            key=lambda node: (
+                max(live_memory + node.mpi_node.size, max_memory),
+                node.mpi_node.size - node.mpi_node.memory_to_free,
+                node.mpi_node.index,
+            ),
+        )
```

Comment:
Could we use a heap here ? instead of sorting on each iteration. for elements which are changed you can remove and reinsert

Reply:
I actually thought about using a heap here but `live_memory` is updating from iteration to iteration, `node.mpi_node.memory_to_free` could be changing from iteration to iteration since how much memory can be freed after scheduling a node depends on what has been sceduled fore.  

So if we were to use a heap all elements in the heap would need to be updated from iteration to iteration

Reply:
I guess the reason this isn't as bad is because it's O(n) * O(max_concurrently_schedulable_nodes).

But, just to discuss - the live memory is the same for every node. If `live_memory + node.mpi_node.size  < max_memory`, does it actually matter what we schedule ? I'm not sure we need the max here: `max(live_memory + node.mpi_node.size, max_memory)`. 

I think you could conceivably avoid updating the live memory and just update the memory_to_free ? anyways not needed but thinking aloud.